spring-cloud-gcp-starters/spring-cloud-gcp-starter-sql/src/main/java/org/springframework/cloud/gcp/sql/autoconfig/GcpCloudSqlAutoConfiguration.java

```diff
@@ -0,0 +1,127 @@
+import org.springframework.boot.autoconfigure.AutoConfigureBefore;
+import org.springframework.boot.autoconfigure.condition.ConditionalOnClass;
+import org.springframework.boot.autoconfigure.condition.ConditionalOnMissingBean;
+import org.springframework.boot.autoconfigure.jdbc.DataSourceAutoConfiguration;
+import org.springframework.boot.context.properties.EnableConfigurationProperties;
+import org.springframework.cloud.gcp.core.GcpProperties;
+import org.springframework.cloud.gcp.sql.GcpCloudSqlProperties;
+import org.springframework.context.annotation.Bean;
+import org.springframework.context.annotation.Configuration;
+import org.springframework.util.Assert;
+import org.springframework.util.StringUtils;
+
+/**
+ * Google Cloud SQL starter.
+ *
+ * <p>It simplifies database configuration in that only an instance and database name are provided,
+ * instead of a JDBC URL. If the instance region isn't specified, this starter attempts to get it
+ * through the Cloud SQL API.
+ *
+ * @author João André Martins
+ */
+@Configuration
+@EnableConfigurationProperties(GcpCloudSqlProperties.class)
+@ConditionalOnClass({GcpProperties.class, GoogleCredential.class})
+@AutoConfigureBefore(DataSourceAutoConfiguration.class)
+public class GcpCloudSqlAutoConfiguration {
+
+	@Autowired
+	private GcpCloudSqlProperties gcpCloudSqlProperties;
+
+	@Autowired
+	private GcpProperties gcpProperties;
+
+	@Bean
+	@ConditionalOnMissingBean(DataSource.class)
+	public DataSource cloudSqlDataSource(SQLAdmin sqlAdmin)
+			throws GeneralSecurityException, IOException {
+		Assert.hasText(this.gcpCloudSqlProperties.getDatabaseName(),
+				"spring.cloud.gcp.sql.databaseName name can't be empty.");
+		Assert.hasText(this.gcpCloudSqlProperties.getInstanceName(),
+				"spring.cloud.gcp.sql.instanceName name can't be empty.");
+		Assert.hasText(this.gcpCloudSqlProperties.getUserName(),
+				"spring.cloud.gcp.sql.userName name can't be null.");
+
+		// Look for the region if the user didn't specify one.
+		String region = this.gcpCloudSqlProperties.getRegion();
+		if (!StringUtils.hasText(region)) {
+			DatabaseInstance instance =	sqlAdmin.instances().get(
+					this.gcpProperties.getProjectId(),
+					this.gcpCloudSqlProperties.getInstanceName()).execute();
+
+			region = instance.getRegion();
+		}
+		String instanceConnectionName = this.gcpProperties.getProjectId() + ":"
+				+ region + ":" + this.gcpCloudSqlProperties.getInstanceName();
+
+		String jdbcUrl = String.format("jdbc:mysql://google/%s?cloudSqlInstance=%s&"
+						+ "socketFactory=com.google.cloud.sql.mysql.SocketFactory",
+				this.gcpCloudSqlProperties.getDatabaseName(),
+				instanceConnectionName);
+
+		// TODO(joaomartins): Set the credential factory to be used by SocketFactory when
+		// https://github.com/GoogleCloudPlatform/cloud-sql-jdbc-socket-factory/issues/41 is
+		// resolved. For now, it uses application default creds.
+
+		HikariConfig hikariConfig = new HikariConfig();
```

Comment:
I feel like this should be a separated out from auto-construction of the `jdbcUrl`. There's too much going on in this method. It would be nice to let the users overwrite which connection pool to use without sacrificing the auto-construction of the JDBC URL.